pydantic/dataclasses.py

```diff
@@ -199,60 +207,124 @@ def _process_class(
-    as Config.validate_assignment.
-    """
-
-    def wrap(cls: Type[Any]) -> Type['Dataclass']:
-        return _process_class(cls, init, repr, eq, order, unsafe_hash, frozen, config)
+# https://github.com/python/mypy/pull/10712
+if TYPE_CHECKING:
+    if sys.version_info >= (3, 10):
+
+        @overload
+        def dataclass(
+            *,
+            init: bool = True,
+            repr: bool = True,
+            eq: bool = True,
+            order: bool = False,
+            unsafe_hash: bool = False,
+            frozen: bool = False,
+            config: Type[Any] = None,
+            kw_only: bool = ...,
+        ) -> Callable[[Type[Any]], Type['Dataclass']]:
+            ...
+
+        @overload
+        def dataclass(
+            _cls: Type[Any],
+            *,
+            init: bool = True,
+            repr: bool = True,
+            eq: bool = True,
+            order: bool = False,
+            unsafe_hash: bool = False,
+            frozen: bool = False,
+            config: Type[Any] = None,
+            kw_only: bool = ...,
+        ) -> Type['Dataclass']:
+            ...
+
+        def dataclass(
+            _cls: Optional[Type[Any]] = None,
+            *,
+            init: bool = True,
+            repr: bool = True,
+            eq: bool = True,
+            order: bool = False,
+            unsafe_hash: bool = False,
+            frozen: bool = False,
+            config: Type[Any] = None,
+            kw_only: bool = False,
+        ) -> Union[Callable[[Type[Any]], Type['Dataclass']], Type['Dataclass']]:
+            ...
 
-    if _cls is None:
-        return wrap
+    else:
 
-    return wrap(_cls)
+        @overload
+        def dataclass(
+            *,
+            init: bool = True,
+            repr: bool = True,
+            eq: bool = True,
+            order: bool = False,
+            unsafe_hash: bool = False,
+            frozen: bool = False,
+            config: Type[Any] = None,
+        ) -> Callable[[Type[Any]], Type['Dataclass']]:
+            ...
+
+        @overload
+        def dataclass(
+            _cls: Type[Any],
+            *,
+            init: bool = True,
+            repr: bool = True,
+            eq: bool = True,
+            order: bool = False,
+            unsafe_hash: bool = False,
+            frozen: bool = False,
+            config: Type[Any] = None,
+        ) -> Type['Dataclass']:
+            ...
+
+        def dataclass(
+            _cls: Optional[Type[Any]] = None,
+            *,
+            init: bool = True,
+            repr: bool = True,
+            eq: bool = True,
+            order: bool = False,
+            unsafe_hash: bool = False,
+            frozen: bool = False,
+            config: Type[Any] = None,
+            kw_only: bool = False,
+        ) -> Union[Callable[[Type[Any]], Type['Dataclass']], Type['Dataclass']]:
+            ...
+
+else:
+
+    def dataclass(
```

Comment:
not sure this should be in `else:`, I think it should be run either with or without `TYPE_CHECKING`?

Reply:
unfortunately i can't due to https://github.com/python/mypy/pull/1071. the workaround requires both sides of the `sys.version_info` check to have a "fake" implementation, so if the real implementation wasn't in the `else` mypy would complain that it's a duplicate definition.

i'll add some more comments to make this more clear

Reply:
See my message above. Just use the same signature for all versions, then raise an error internally.

Should solve most of these problems.

Reply:
but that way you only get a runtime error. imo it's better to allow type checkers to catch errors like these wherever possible.

if you are happy with this approach, i've made a change to the mypy tests to allow version specific output files which i believe solves that problem

Reply:
https://github.com/python/mypy/pull/10712 has now been merged which solves the issue